src/lib/udev/google_nvme_id

```diff
@@ -0,0 +1,305 @@
+  return 0
+}
+
+#######################################
+# Retrieves the partition number for a device path if it exists
+# Globals:
+#   None
+# Arguments:
+#   The path to the device partition (/dev/nvme0n*p*)
+# Outputs:
+#   The value after 'p' in the device path, or an empty string if the path has
+#   no partition.
+#######################################
+function get_partition_number() {
+  local dev_path="${1}"
+  local partition_number
+  if [[ "${dev_path}" =~ $PARTITION_NUMBER_REGEX ]]; then
+    partition_number="${BASH_REMATCH[1]}"
+    echo "${partition_number}"
+  else
+    echo ''
+  fi
+  return 0
+}
+
+#######################################
+# Generates a symlink for a PD-NVMe device using the metadata's disk name.
+# Primarily used for testing but can be used if the script is directly invoked.
+# Globals:
+#   Uses ID_SERIAL_SHORT (can be populated by identify_pd_disk)
+# Arguments:
+#   The device path for the disk
+#######################################
+function gen_symlink() {
+  local dev_path="${1}"
+  local partition_number="$(get_partition_number "${dev_path}")"
+
+  if [[ -n "${partition_number}" ]]; then
+    ln -s "${dev_path}" /dev/disk/by-id/google-persistent-disk-"${ID_SERIAL_SHORT}"-part"${partition_number}" > /dev/null 2>&1
+  else
+    ln -s "${dev_path}" /dev/disk/by-id/google-persistent-disk-"${ID_SERIAL_SHORT}" > /dev/null 2>&1
+  fi
+
+  return 0
+}
+
+#######################################
+# Validates that a given path corresponds to a valid NVMe device
+# Globals:
+#   Uses nvme_cli_bin
+# Arguments:
+#   The device path for the disk
+# Returns:
+#   The return value of the call to the nvme id-ctrl command
+#######################################
+function validate_nvme_device() {
+  "${nvme_cli_bin}" id-ctrl "${1}" &>/dev/null
+}
+
+#######################################
+# Determines if the device path is a PD or Local SSD disk
+# Globals:
+#   Uses nvme_cli_bin, PD, and LOCAL_SSD
+# Arguments:
+#   The device path for the disk
+# Returns:
+#   1 if the device type is Unknown, 0 on success
+#######################################
+function detect_pd_or_local_ssd() {
+  local nvme_id_ctrl
+  nvme_id_ctrl="$("${nvme_cli_bin}" id-ctrl "${1}")"
+  if [[ "${nvme_id_ctrl}" =~ 'nvme_card-pd' ]]; then
+    echo "${PD}"
+  elif [[ "${nvme_id_ctrl}" =~ 'nvme_card' ]]; then
+    echo "${LOCAL_SSD}"
+  else
+    return 1
+  fi
+  return 0
+}
+
+#######################################
+# Populates the ID_* global variables with a disk's device name and namespace
+# Globals:
+#   Populates ID_NAMESPACE_INDEX, ID_SERIAL_SHORT, and ID_SERIAL
+# Arguments:
+#   The device path for the disk
+# Returns:
+#   0 on success and 255 if an error occurrs
+#######################################
+function identify_pd_disk() {
+  local dev_path="${1}"
+  local dev_name
+  dev_name="$(get_namespace_device_name "${dev_path}")"
+  if [[ $? != 0 ]]; then
+    return 255
+  fi
+
+  ID_NAMESPACE_INDEX="${namespace_index}"
```

Comment:
where does $namespace_index come from? where is $ID_NAMESPACE_INDEX used?

Reply:
also note, you are setting globals as a side effect of invoking the function, rather than using return values or outputs. that's hard to follow, if you can easily avoid it you should

Reply:
Hold over from a previous version of the code.  W.r.t the globals: it makes sense to me to keep those as globals since they're the primary output of the script, and it would be hard and a bit kludgy to split them out into their own return values.  I've added a comment to their declaration clarifying where they're set and why.